### PR TITLE
Fix loading of submodules

### PR DESCRIPTION
**Pull request by rhaschke**
Fixup suggested in https://github.com/ros-planning/moveit/pull/2547#discussion_r596207428.

### CODE REVIEW

**Comment by PeterMitrano** [comment]
##  tl;dr

I don't think we should do this, but if we do, we need to include `from pymoveit_core import load_robot_model` instead of just deleting the first line.

## full answer

So I agree what you test example does not work, but I don't think it should, and I think your proposed change isn't quite complete.

I would not expect to be able to write
```
import moveit_core
moveit_core.robot_model
```
because robot_model is a submodule, not a file or symbol. Just like you can't write this
```
import scipy
scipy.stats
```

However, you can write `from moveit_core import robot_model` or `import moveit_core.robot_model` and that works fine. This behavior seems to be intentional from the python team:

> The behavior of other modules is intentional.
> In general it is not a good idea to automatically import all submodules
when you import a top-level package; imaging what this would do if you have
a large top-level package (maybe a namespace package) containing hundreds
of modules.
> For specific top-level packages that want to provide all (or some)
submodules upon import of just the top-level package, they should have the
sub-imports in the top-level __init__.py.
> --Guido van Rossum (python.org/~guido)

Of course, we _can_ add the line you suggest, which does automatically import submodules, but I don't think we need to.

But we can't simply remove the first line, because then symbols defined for pymoveit_core are not available. For right now the only one is `load_robot_model` but still. So as is we would need to replace `from pymoveit_core import *` with `from pymoveit_core import load_robot_model, ...`. That would be an ok compromise, I don't think we'll be adding many functions there.

**Comment by rhaschke** [comment]
> We need to include `from pymoveit_core import load_robot_model` instead of just deleting the first line.

Yep, I agree. I missed the dangling `load_robot_model` symbol.

> For specific top-level packages that want to provide all (or some)
> submodules upon import of just the top-level package, they should have the
> sub-imports in the top-level **init**.py.

Here the python team explicitly suggests this use case - if desired.
Actually, by doing `from pymoveit_core import *` you allowed this use case as well.
What _was_ missing, was the possibility to augment and override the C++ lib modules/symbols.

File: moveit_core/python/src/moveit/core/__init__.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# load symbols from C++ extension lib
+from pymoveit_core import *
+
+# and augment with symbols from python modules (order is important to allow overriding!)
```
Comment (PeterMitrano) [inline moveit_core/python/src/moveit/core/__init__.py:4]: that's a little spooky. why does order matter?

File: moveit_core/setup.py [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
 
 from catkin_pkg.python_setup import generate_distutils_setup
-from setuptools import find_packages, setup
+from setuptools import setup
 
 d = generate_distutils_setup(
-    packages=find_packages('src'),
-    package_dir={'': 'src'}
+    packages=["moveit"],
+    package_dir={"": "python/src"}
```
Comment (PeterMitrano) [inline moveit_core/setup.py:8]: did you check if this still works correctly when doing catkin install? I thought you had to list all modules & submodules explicitly for it to work, hence the `find_packages`

File: moveit_core/python/src/moveit/core/__init__.py (hunk 1 above)
Comment (rhaschke) [reply moveit_core/python/src/moveit/core/__init__.py:4]: Because the python code might want to override what was previously defined in the C++ extension lib. Much like in the following python code, only the second definition "survives":
```python
def foo():
    print("foo")
def foo():
    print("bar")
```
We used this pattern heavily in the past in MoveIt's python wrappers.

File: moveit_core/setup.py (hunk 2 above)
Comment (rhaschke) [reply moveit_core/setup.py:8]: Maybe, you are right. I used this pattern successfully for MTC, but I'm not sure I tested with install space.
The more pressing issue is that we actually don't want to install a `moveit` package here, but only the sub-package `core` into the `moveit` package. I'm not yet sure how to do this...

Comment (PeterMitrano) [reply moveit_core/setup.py:8]: this is what I was talking about in the thread on the other PR. I think to do one-big moveit we'd have to move some of this stuff up a level in terms of directory structure. Maybe we can put it in the empty moveit (meta?)package? The setup.py as well as any top-level C++ bindings and pure python code living directly in the `moveit` namespace and not under any sub-namespace could live there. That's just an idea though I haven't looked into implementing that

File: moveit_core/python/src/moveit/core/__init__.py (hunk 1 above)
Comment (PeterMitrano) [reply moveit_core/python/src/moveit/core/__init__.py:4]: oh yikes I don't love that :grimacing: can you point me to an example? I'm sure you had a good reason for doing it, but I can't find any examples in moveit_commander.

Comment (PeterMitrano) [reply moveit_core/python/src/moveit/core/__init__.py:4]: of course we can keep the order as you suggest, it does no harm. So I'll incorporate this into the PR once we figure out the other restructuring stuff we're talking about

Comment (rhaschke) [reply moveit_core/python/src/moveit/core/__init__.py:4]: > can you point me to an example?

`moveit_commander` doesn't do it exactly like this, but very similar. First the C++ extension lib is loaded:
https://github.com/UM-ARM-Lab/moveit/blob/d18173b598846260d4e067ff604c3f3b29366f57/moveit_commander/src/moveit_commander/planning_scene_interface.py#L40
and then its classes (with exactly the same name) are wrapped by python code:
https://github.com/UM-ARM-Lab/moveit/blob/d18173b598846260d4e067ff604c3f3b29366f57/moveit_commander/src/moveit_commander/planning_scene_interface.py#L57
https://github.com/UM-ARM-Lab/moveit/blob/d18173b598846260d4e067ff604c3f3b29366f57/moveit_commander/src/moveit_commander/planning_scene_interface.py#L65-L66

> oh yikes I don't love that

I admit that _overriding_ loaded symbols usually will not make sense. :wink:

Comment (PeterMitrano) [reply moveit_core/python/src/moveit/core/__init__.py:4]: Ok yea that makes much more sense!

File: moveit_core/setup.py (hunk 2 above)
Comment (PeterMitrano) [reply moveit_core/setup.py:8]: Ok I just updated things (check the PR) -- I think I have it working for the use cases we've considered, and I tested by making bindings for moveit_ros_occupancy_map_updater. They can be installed alongside moveit/core without any problems from what I can tell :crossed_fingers:

**Comment by PeterMitrano** [comment]
I'm going to close this since I've incorporated the chances, but you can reopen if you want
